pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -450,70 +489,88 @@ func (c *Controller) onServiceEvent(curr interface{}, event model.Event) error {
 
 	log.Debugf("Handle event %s for service %s in namespace %s", event, svc.Name, svc.Namespace)
 
-	svcConv := kube.ConvertService(*svc, c.opts.DomainSuffix, c.Cluster(),
-		c.imports.GetClusterSetIPs(kube.NamespacedNameForK8sObject(svc)))
+	// Create the standard (cluster.local) service.
+	svcConv := kube.ConvertService(*svc, c.opts.DomainSuffix, c.Cluster())
 	switch event {
 	case model.EventDelete:
-		c.Lock()
-		delete(c.servicesMap, svcConv.ClusterLocal.Hostname)
-		delete(c.nodeSelectorsForServices, svcConv.ClusterLocal.Hostname)
-		delete(c.externalNameSvcInstanceMap, svcConv.ClusterLocal.Hostname)
-		_, isNetworkGateway := c.networkGateways[svcConv.ClusterLocal.Hostname]
-		delete(c.networkGateways, svcConv.ClusterLocal.Hostname)
-		c.Unlock()
-		if isNetworkGateway {
-			// networks are different, we need to update all eds endpoints
-			c.opts.XDSUpdater.ConfigUpdate(&model.PushRequest{Full: true, Reason: []model.TriggerReason{model.NetworksTrigger}})
-		}
+		c.deleteService(svcConv)
 	default:
-		needsFullPush := false
-		// First, process nodePort gateway service, whose externalIPs specified
-		// and loadbalancer gateway service
-		if !svcConv.Attributes.ClusterExternalAddresses.IsEmpty() {
-			needsFullPush = c.extractGatewaysFromService(svcConv)
-		} else if isNodePortGatewayService(svc) {
-			// We need to know which services are using node selectors because during node events,
-			// we have to update all the node port services accordingly.
-			nodeSelector := getNodeSelectorsForService(svc)
-			c.Lock()
-			// only add when it is nodePort gateway service
-			c.nodeSelectorsForServices[svcConv.ClusterLocal.Hostname] = nodeSelector
-			c.Unlock()
-			needsFullPush = c.updateServiceNodePortAddresses(svcConv)
-		}
+		c.addOrUpdateService(svc, svcConv, event)
+	}
 
-		// instance conversion is only required when service is added/updated.
-		instances := kube.ExternalNameServiceInstances(svc, svcConv)
+	return nil
```

Comment:
very nit: seems like 
```
	c.opts.XDSUpdater.SvcUpdate(shard, string(svcConv.ClusterLocal.Hostname), svc.Namespace, event)
	// Notify service handlers.
	for _, f := range c.serviceHandlers {
		f(svcConv, event)
	}
``

could remain here since its the same for all events

Reply:
I copied to each of the separate methods so that I could call them independently without calling `onServiceEvent` manually, which seems icky.